[REVIEW] Add filter tests and nested type host side API [skip ci]

kuhushukla opened the pull request:
Fixes a small bug and adds java test for filter on structs provided via https://github.com/rapidsai/cudf/pull/6271 which is underway , hence this is WIP as I want to add to this PR - more tests and functionality/correctness checks.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -3483,9 +3483,8 @@ private static NestedColumnVector createNestedColumnVector(DType type, long rows
-        } else if (type.isNestedType()) {
-          dataLen = dataBuffer.length;
         }
+        dataLen = Math.min(dataLen,dataBuffer.length);
```
What is the situation that caused this change?  This indicates that either someone is not giving us as much data as we expected, or our calculation is off.  Nested data types are the ones not truly covered here, and in what situation do we have a list, or a struct with an actual data column in it?

kuhushukla replied:
The issue is `long dataLen = rows * type.sizeInBytes;` does not consider nulls, so we expect to copy more than what is actually there. Agree this is not really nested type related, in fact when we make the child vectors of base type with nulls, this becomes a problem. As I understand, we want to write values except the nulls.

revans2 replied:
So this is for a child column of a list?  If so we need to have the correct number of rows for that child column, not for any parent column.

kuhushukla replied:
I thought I had that working for rows of nested child column , I will check and fix. Thanks for the help!

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -107,10 +107,11 @@ public static void assertPartialStructColumnsAreEqual(ColumnVector expected, lon
   public static void assertPartialColumnsAreEqual(ColumnVector expected, long rowOffset, long length,
-                                                  ColumnVector cv, String colName, boolean enableNullCheck) {
+                                                  ColumnVector cv, String colName, boolean enableNullCheck,
+                                                  HostColumnVector.ColumnBuilder.DataType type) {
```
Can we please make `type` a little clearer?  In other places it is called schema.  Also at this point it might be good to have some javadocs to explain what is happening here.  Honestly we should have done it a lot sooner.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -291,14 +292,33 @@ public static void assertPartialTablesAreEqual(Table expected, long rowOffset, l
+  }
+
+  public static void assertPartialTablesAreEqual(Table expected, long rowOffset, long length, Table table,
+                                                 boolean enableNullCheck, HostColumnVector.ColumnBuilder.DataType type) {
```
type does not work here generically.  For this particular API call type is the type of any and all structs that are in the table.  That is not maintainable.  There can be lots of tests with multiple structs in them, or nested structs.  We should either be able to determine the schema of the data from the column itself, or we should have this take in the schema for the entire table, not just a single column.

kuhushukla replied:
That is true :( Sorry I missed that, Let me fix this!

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -3386,4 +3406,56 @@ void testORCWriteToFileUncompressed() throws IOException {
+    type.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 4, DType.INT64));
+    HostColumnVector.ColumnBuilder.StructType expectedType = new HostColumnVector.ColumnBuilder.StructType(true, 2);
+    expectedType.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 2, DType.INT32));
+    expectedType.addChild(new HostColumnVector.ColumnBuilder.BasicType(true, 2, DType.INT64));
```
Wow this looks painful to write.  Can we please have a follow on issue to clean this up?  When building up test data we should just need to provide the data, and some hint of what the type of the data should be.

kuhushukla replied:
I will start a follow on if I cant fix it here. I know this is not great :(

kuhushukla commented:
Pushed some changes here to the test API which does not take premeditated rows anymore. Could use more direction if this is not exactly what want. I have also kept the data buffer as long as the `expected size`. Added a table schema which is alist of datatypes for testing. Should we remove `nullable` from DataType since it is not really used? I moved the DataType and associated classes out the builder but if that is not preferred, I can move them back.

kuhushukla commented:
Will update here with more tests just to make sure data looks right etc.

kuhushukla commented:
The daya looks ok for a couple tests in TableTest and ColumnVectorTest and the asserts seem to do that anyway. 
> Pushed some changes here to the test API which does not take premeditated rows anymore. Could use more direction if this is not exactly what want. I have also kept the data buffer as long as the expected size. Added a table schema which is alist of datatypes for testing. Should we remove nullable from DataType since it is not really used? I moved the DataType and associated classes out the builder but if that is not preferred, I can move them back.
>
@revans2 could use your inputs on the changes and what all we can do as part of this change for APIs required.

revans2 reviewed (approved):
This is definitely a step more in the right direction.

kuhushukla commented:
@revans2 @mythrocks I upmerged and reran the build and java tests and they pass. Marking this out of WIP and merging with one approval.
